File: pearls/anagrams/anagram.hs
import qualified Data.Map as Map
import Data.Ord
import Data.List

anagrams words = 
  sorted_anagrams
  where  
    sorted_anagrams      = sortBy (flip $ comparing length) get_anagrams
    get_anagrams         = Map.elems $ foldl' insert_word Map.empty words
    insert_word map word = Map.insertWith' (++) (sort word) [word] map

main = do
  input <- getContents
  print $ take 3 $ anagrams $ lines input

File: pearls/anagrams/anagram.rb
def get_anagrams words
  result = Hash.new []
  words.each do |word|
    word.chomp!
    sorted_word = word.chars.sort.join
    result[sorted_word] += [word]
  end
  result.values.sort! { |a, b| b.size <=> a.size }
end

input = STDIN.readlines

p get_anagrams(input)[0..2]
